Misaligned Vectorization Support

rdspring1 opened the pull request:
- [X] Get full extent for the inner-most, merged root domain
- [x] Add predicate for all root domains except vectorized
- [x] Global Memory Check
- [x] Check that input and output tensors have the same shift and/or stride

Misaligned Vectorization pseudo-code for Point-wise Addition
```
__global__ void CUDAGeneratedKernel(Tensor<float, 4> T0, Tensor<float, 4> T1, Tensor<float, 4> T2) {
  #pragma unroll 1
  for(size_t ki121 = 0; ki121 < T0.size[0]; ++ki121) {
    #pragma unroll 1
    for(size_t ki122 = 0; ki122 < (ceilDiv((T0.size[2] * T0.size[3]), 256)); ++ki122) {
      // Allocations
      float T4[4];
      float T3[4];
      // Alias Allocation - register
      float* T5 = T3;

      if ((((blockIdx.x * 128) + blockIdx.y) < T1.size[1])) {
        int64_t base_address;
        base_address = (...);

        int64_t shift;
        shift = (((ceilDiv(base_address, 4)) * 4) - base_address);

        int64_t remainder;
        remainder = (((T1.size[3] * T1.size[2]) - shift) % 4);

        int64_t threshold;
        threshold = (((T1.size[3] * T1.size[2]) - 4) - shift);

        int64_t last_root_dim_index;
        last_root_dim_index = (((ki122 * ((ceilDiv(256, 4)) * 4)) + ((threadIdx.x * 4) + 0)) * 1);

        // Vectorization
        if ((last_root_dim_index <= threshold)) {
          *reinterpret_cast<Array<float, 4>*>(&T4[0]) 
                            = *reinterpret_cast<Array<float, 4>*>(&T1[index + shift)]);
          *reinterpret_cast<Array<float, 4>*>(&T3[0]) 
                            = *reinterpret_cast<Array<float, 4>*>(&T0[index + shift)]);
          #pragma unroll 1
          for(size_t ki128 = 0; ki128 < 4; ++ki128) {
            T5[ki128]
              = T3[ki128]
              + T4[ki128];
          }
          *reinterpret_cast<Array<float, 4>*>(&T2[index + shift)]) 
                            = *reinterpret_cast<Array<float, 4>*>(&T5[0]);
        }

        // Pre - Shift
        if ((last_root_dim_index == 0)) {
          #pragma unroll 1
          for(size_t ki134 = 0; ki134 < shift; ++ki134) {
            T4[ki134]
               = T1[index];
          }
          #pragma unroll 1
          for(size_t ki133 = 0; ki133 < shift; ++ki133) {
            T3[ki133]
               = T0[index];
          }
          #pragma unroll 1
          for(size_t ki128 = 0; ki128 < shift; ++ki128) {
            T5[ki128]
              = T3[ki128]
              + T4[ki128];
          }
          #pragma unroll 1
          for(size_t ki125 = 0; ki125 < shift; ++ki125) {
            T2[index]
               = T5[ki125];
          }
        }

        // Post - Remainder
        if (((last_root_dim_index > threshold) && (last_root_dim_index < ((T1.size[3] * T1.size[2]) - shift)))) {
          #pragma unroll 1
          for(size_t ki134 = 0; ki134 < remainder; ++ki134) {
            T4[ki134]
               = T1[index + shift)];
          }
          #pragma unroll 1
          for(size_t ki133 = 0; ki133 < remainder; ++ki133) {
            T3[ki133]
               = T0[index + shift)];
          }
          #pragma unroll 1
          for(size_t ki128 = 0; ki128 < remainder; ++ki128) {
            T5[ki128]
              = T3[ki128]
              + T4[ki128];
          }
          #pragma unroll 1
          for(size_t ki125 = 0; ki125 < remainder; ++ki125) {
            T2[index + shift)]
               = T5[ki125];
          }
        }
      }
    }
  }
}
```

rdspring1 committed:
Create Misaligned Vectorization Structure

rdspring1 committed:
Add extent field to the For-Loop

rdspring1 committed:
Add vectorize field to For-Loop

rdspring1 committed:
Update Test

rdspring1 committed:
Enable Vectorized Indexing for misaligned vectorization

rdspring1 committed:
Use vectorize index as base_address

last_dim_index = exclude stride and last iterDomain offset

rdspring1 committed:
Add lshift offset to global vectorize indices

Fix lshift value - ceilDiv(x,4)*4 - x

Replace floorDiv(N, 4) with (N - lshift)

Disable aliasing for vectorized allocations to fix correctness

rdspring1 committed:
Fix vectorize scope

rdspring1 committed:
Fix If-Statement predicates

rdspring1 committed:
clang-tidy - const correctness

rdspring1 committed:
Replace Inline expression with NamedScalar value

Update FusionParser test

rdspring1 committed:
Add upper-bound threshold for the remainder phase

Add rFactor Misaligned Vectorization test

rdspring1 committed:
Add basic validation

rdspring1 committed:
Re-enable aliasing for vectorize TensorViews

rdspring1 committed:
Get full extent for the inner-most, merged root domain

rdspring1 committed:
Add predicate for all root domains except the misaligned vectorized domain

naoyam commented on torch/csrc/jit/codegen/cuda/lower_validation.cpp:
```diff
@@ -247,6 +248,14 @@ void validateVectorize(Fusion* fusion) {
         has_vectorize_dim = true;
       }
+
+      if (concrete_id->getParallelType() == ParallelType::MisalignedVectorize) {
+        TORCH_INTERNAL_ASSERT(
+            !tv->hasComputeAt() ||
+                tv->getComputeAtPosition() == tv->nDims() - 1,
+            "Only allow misaligned vectorization in the -2 computeAt position.");
```
Where do we validate the vectorized axis is inner-most?

rdspring1 replied:
`validateVectorizedTensors` in executor_utils and `validateVectorize` in lower_validation

naoyam replied:
I'm trying to find where it is for `MisalignedVectorize`. I think `validateVectorizedTensors` only looks at `Vectorize`, which makes sense. Where should I look at in `validateVectorize` for `MisalignedVectorize`?

rdspring1 replied:
It is in the same spot. 

https://github.com/csarofeen/pytorch/blob/dfc1690a7f0779a9111c7279e757414b6bee0d86/torch/csrc/jit/codegen/cuda/lower_validation.cpp#L119-L212

naoyam replied:
Thanks! I found it's in the very end of that function.

naoyam commented on torch/csrc/jit/codegen/cuda/kernel_ir.h:
```diff
@@ -1244,9 +1269,12 @@ class TORCH_CUDA_CU_API ForLoop final : public Expr {
   Scope body_;
   bool unroll_ = false;
+  bool vectorize_ = false;
+  Val* shift_ = nullptr;
```
Is `shift_` supposed to be used only when `vectorize` is true?

rdspring1 replied:
`[pre | vectorize | post]`  <= inner-most, merged root domain
`shift_` is applied to the `vectorize` and `post` sections.

naoyam replied:
When is `vectorize_` true? Please comment them.

rdspring1 replied:
// vectorize is true when the for-loop contains a vectorize set
  // the flag is used to omit the for-loop from the kernel

naoyam commented on torch/csrc/jit/codegen/cuda/kernel_ir.h:
```diff
@@ -1244,9 +1269,12 @@ class TORCH_CUDA_CU_API ForLoop final : public Expr {
 
  private:
   Val* const index_ = nullptr;
+  Val* const extent_ = nullptr;
```
Should `extent_` always be used instead of `inter_domain_->extent()`? I see both are used and don't know that's intentional.

rdspring1 replied:
`extent_` is primarily used in the codegen. `for(int idx = init; idx < extent; idx += offset)`

naoyam replied:
Even in codegen.cpp, `ForLoop::iter_domain()->rawExtent()` is used. I think it's confusing to have both `ForLoop::extent()` and `ForLoop::iter_domain()->extent()`.

I see the only place where `ForLoop::extent()` can be different from `ForLoop::iter_domain()->extent()` is `cloneVectorizedLoopNests`. I wonder if it would make more sense to create a new `kir::IterDomain` rather than reusing the same `kir::IterDomain` for the cloned loops. That way, we don't need to have `ForLoop::extent()` to override the extent of `ForLoop::iter_domain()`.

rdspring1 replied:
@csarofeen Can we create an new iterDomain for the different loops?

I forgot if this idea was thrown around before.

naoyam replied:
I couldn't find a clean way to create new IterDomains during lowering. I'll try to make it a little less confusing. Will create a PR when ready.

naoyam replied:
Please review #784

csarofeen replied:
Seemed reasonable but left one comment on the PR.

naoyam commented on torch/csrc/jit/codegen/cuda/index_compute.cpp:
```diff
@@ -914,7 +930,10 @@ std::unordered_map<kir::ForLoop*, kir::Val*> indexMapFromTV(
+        (loop->iter_domain()->parallelType() == ParallelType::Vectorize) ||
+        (loop->iter_domain()->parallelType() ==
+             ParallelType::MisalignedVectorize &&
+         loop->vectorize())) {
```
Why do we need to check all three of them? Isn't `loop->vectorize()` enough?

rdspring1 replied:
Yes

naoyam commented on torch/csrc/jit/codegen/cuda/kernel_ir.h:
```diff
@@ -1011,13 +1012,18 @@ class TORCH_CUDA_CU_API TensorIndex final : public Val {
  private:
   const TensorView* view_ = nullptr;
   std::vector<Val*> indices_;
+  std::vector<bool> root_domain_index_map_;
```
What does this bool vector mean? Can you please add a comment?

rdspring1 replied:
I need to find the full size of the inner-most, merged root domain.
A root domain has a zero index when it is merged with another domain.
I use this bool vector to mark which root domain indices are zero.

Is there a more robust way to find which root domains are merged together?

naoyam commented:
Does it make sense to use `MisalignedVectorize` with non-global memory tensors? If not, shouldn't the validation check that as well?

naoyam commented on torch/csrc/jit/codegen/cuda/index_reference_replay.cpp:
```diff
@@ -261,8 +265,11 @@ IndexCompute getReferenceIndexing(
+        (loop_structure[loop_i]->iter_domain()->parallelType() ==
+             ParallelType::Vectorize ||
+         loop_structure[loop_i]->iter_domain()->parallelType() ==
+             ParallelType::MisalignedVectorize)) {
```
When `vectorize()` is false but `iter_domain()->parallelType()` is either `Vectorize` or `MisalignedVectorize`?

rdspring1 replied:
It isn't.

naoyam commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -109,6 +324,19 @@ void UnrollPass::handle(kir::Expr* expr) {
   }
 }
 
+bool containsMisalignedVectorization(const kir::ForLoop* fl) {
```
nit: move this to an anonymous namespace

naoyam commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,221 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+    const std::vector<kir::ForLoop*>& for_loops) {
+  for (auto fl : for_loops) {
+    auto first_expr = fl->body().exprs().front();
+    bool has_vectorize_op =
```
Does this mean that vectorization only happens the first expression is a UnaryOp::Set? What happens if other expressions exist in the same loop?

rdspring1 replied:
No. We are using the first vectorized set op to get the vectorized index. The assumption is that the shift offset is the same for all the global tensors. We can check this assumption in `validateVectorizedTensors`.

csarofeen replied:
I don't understand this assumption, can you elaborate?

naoyam commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,221 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+    bool has_vectorize_op =
+        (first_expr->isA<kir::UnaryOp>() &&
+         first_expr->as<kir::UnaryOp>()->operation() == UnaryOpType::Set);
+    if (has_vectorize_op) {
```
Don't we need to check `fl->iter_domain()->parallelType()` is `MisalignedVectorize`?

rdspring1 replied:
👍🏼

naoyam commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,221 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+  TORCH_INTERNAL_ASSERT(vec_expr != nullptr);
+
+  // out_tv is the TensorView with the misaligned vec iterDomain
+  auto out_tv = vec_expr->outputs().front()->as<kir::TensorView>();
```
Are we certain `vec_expr->outputs().front()` is always `kir::TensorView`? I don't find any explicit check.

rdspring1 replied:
I added a check.

naoyam commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,221 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+  auto last_root_dim_index = setupNamedScalar(
+      pred_ite->thenBody(), index->indices().back(), "last_root_dim_index");
+
+  // Part A - Vectorize
```
Is there any reason to do the vectorized part first? I'd imagine B->A->C would be more cache friendly in general, but maybe there's some other things to consider for GPUs.

rdspring1 replied:
Not really. Christian likes having the main part first, then the clean up sections.

naoyam replied:
@csarofeen Is that preferred on GPUs than just linearly accessing?

rdspring1 replied:
I thought it was a stylistic choice.

naoyam replied:
OK, now I understand this is really just stylistic. I thought this ordering would mean we would first do the vectorized part and then the pre and post parts, which isn't actually the case, but they are just placed there in that order.

csarofeen replied:
Unclear if there would be any perf impacts. Vectorized reads is like concern #1 for memory access, so we're already >90% of the way there.

More complex uses there could be a couple effects, yes memory cache is one, instruction cache could be another. These might be at odds for which order we do these sections in.

Stylistically it's nice to see the vectorized section first in the generated code, and the cleanup after. Which order they're executed in is another conversation and hopefully the implementation doesn't firmly lock us into one choice.

naoyam commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,221 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+  auto root_domain_index_map = node->rootDomainIndexMap();
+
+  auto extent = domain.back()->extent();
+  for (int64_t i = domain.size() - 2; i >= 0; --i) {
```
Let's make sure `domain.size() >= 2` before doing the subtraction since `size()` returns unsigned integer. Otherwise, subtracting it with 2 could result in an unintended value.

rdspring1 committed:
General fixes

naoyam commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,221 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+  return nullptr;
+}
+
+// Get full extent for the inner-most, merged root domain
```
I must be missing something here, but why do we need to look at the root domain? Why not just `node->view()->axis(-1)->extent()` ?

naoyam commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,228 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+  auto out_tv = vec_expr->outputs().front()->as<kir::TensorView>();
+  auto in_tv = vec_expr->inputs().front()->as<kir::TensorView>();
+  kir::TensorView* vec_tv =
+      (out_tv->memoryType() == MemoryType::Local) ? out_tv : in_tv;
```
I'm guess here that you are trying to find whether this should be vectorized store or load, and if it's store, the producer tensor must be on local memory, and if it's load, the consumer must be local memory.

What if we want to do vectorized-load to shared memory? What about vectorized store from shared memory?

rdspring1 replied:
Here I'm looking for the TensorView with the Misaligned Vectorization axis.

`out_tv->memoryType() != MemoryType::Global` is a better choice.

naoyam replied:
I've seen vectorized load from shared to registers. 

Let's focus on global <-> register vectorization for now. Just please make sure erroring out for other cases.

rdspring1 replied:
How do you propose to handle the shift (misalignment) in shared memory? The shared memory access need to be aligned as well.

naoyam replied:
Vectorized load from shared to local should look similar to vectorized load from global to local. 

Vectorized load from global to shared seems tricky. We should probably just use the memcpy_async intrinsic.

csarofeen replied:
Vectorized loads from smem are most urgently needed aligned. Unclear if there are any use cases of misaligned for smem.

The only reason we do vectorized loads from smem in practice is for gemm/conv which uses aligned vectorization.

Might be some uses some day, but definitely not an urgent concern.

naoyam commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,228 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+  new_loop->body().push_back(pred_ite);
+
+  // Generate vectorize index
+  auto index = (out_tv->memoryType() == MemoryType::Global)
```
Similarly here, what if we want to use shared memory as well? 

I think it's fine to not support it for now. Let's make sure that's validated somewhere (sorry if it's actually done somehwhere).

rdspring1 replied:
Here we need the global memory index to generate the `shift` offset.

Adding more validation is on my TO-DO list.

I'm not sure if we can support global <-> shared misaligned vectorization.

naoyam commented:
EDIT: This was just my misunderstanding. Please ignore it.

In the above example, for the vectorized part, don't you need to check whether the last idx is greater or equal to `shift`?

Specifically, here:
```
 // Vectorization
        if ((last_root_dim_index <= threshold)) 
```

Shouldn't it be like below?:
```
 // Vectorization
        if ((last_root_dim_index >= shift && last_root_dim_index <= threshold)) 
```

rdspring1 commented:
The 0-index threshold handles the pre-section and the first vectorized access at +shift.

naoyam commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,228 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+  auto a = ir_builder.ceilDivExpr(base_address, vector_size);
+  auto b = ir_builder.mulExpr(a, vector_size);
+  auto shift_val = ir_builder.subExpr(b, base_address);
+  auto shift = setupNamedScalar(pred_ite->thenBody(), shift_val, "shift");
```
I'm still confused here. This is the corresponding part from the generated code of the MisalignedPointwise test:

```
      int64_t base_address;
      base_address = ((blockIdx.x * (1 * T1.size[1])) + (((ki70 * ((ceilDiv(256, 4)) * 4)) + ((threadIdx.x * 4) + 0)) * 1));
      int64_t shift;
      shift = (((ceilDiv(base_address, 4)) * 4) - base_address);
```

Isn't this, `(((ki70 * ((ceilDiv(256, 4)) * 4)) + ((threadIdx.x * 4) + 0)) * 1)`, always divisible by `vector_size`? If so, `base_address` is always divisible for `blockIdx.x * (1 * T1.size[1])` is divisible by `vector_size`. So, for example, thread blocks with `blockIdx.x == 0`, `base_address` is always divisible by `vector_size`, so `shift` is always zero for the thread blocks. Does this really address misaligned accesses? I wonder if the actual address of `vec_tv` needs to be considered.

rdspring1 replied:
Base Address isn't always divisible be vector size.

If blockIdx.x == 1 and the tensor is [32, 47], is that address aligned.

It is 47 % 4 == 3.

If blockIdx.x == 2, then it is 94 % 4 == 2.

The first index is always aligned.

rdspring1 replied:
Why didn't you consider a blockIdx.x > 0?

rdspring1 committed:
Use pointer address instead of index

Modify TensorIndex to generate pointer address

naoyam commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,222 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+      setupNamedScalar(pred_ite->thenBody(), address, "base_address");
+
+  // NOLINTNEXTLINE(cppcoreguidelines-avoid-magic-numbers)
+  kir::Int* sixteen = ir_builder.create<kir::Int>(16);
```
Is this sixteen because we want to find 16-byte boundaries?

rdspring1 replied:
I replaced the constants with dynamic values.

naoyam commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,222 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+  kir::Int* sixteen = ir_builder.create<kir::Int>(16);
+  kir::Int* four = ir_builder.create<kir::Int>(4);
+  auto a = ir_builder.modExpr(base_address, sixteen);
+  auto b = ir_builder.divExpr(a, four);
```
Why divide by 4?

rdspring1 committed:
Print only Then-block if predicate is always true

rdspring1 committed:
Check that input and output tensors have the same stride

Global Memory Check

naoyam commented on torch/csrc/jit/codegen/cuda/kernel_ir.h:
```diff
@@ -1011,13 +1013,27 @@ class TORCH_CUDA_CU_API TensorIndex final : public Val {
+  std::vector<bool> root_domain_index_map_;
+
+  // Generate the address for the TensorIndex
+  bool address_;
```
This seems like a hack. Why not adding `UnaryOpType::AddressOf`? That way, we don't need to add this flag. Also, it should be possible to use the new unary op with any val, so it should be more generic.

rdspring1 replied:
I modified `TensorIndex` based on Christian's suggestion. He avoided adding a new `UnaryOpType` before because he didn't want them exposed to the user. e.g., `UnaryOpType::VectorizedSet` You can ask him which he prefers in this case.

naoyam replied:
@csarofeen

That doesn't seem to be a concern since it's only for KIR. It's not possible to hide it from the user since the enum is shared between the two IRs, but we can error out when used in user programs.

It's fairly natural that some ops are only supported in either of the IRs, so if we really want to hide unsupported ops, we should define a separate enum for each IR.

`&` is a standard unary op in CUDA, and I don't see any reason we should not have that in KIR.

csarofeen replied:
The unary ops I really didn't like were the vectorized read and vectorized write one. I agree this unary op is a bit of a concern, but it does seem to make sense to have in the kernel ir. Since we don't have a separate enum for the kernel ir, what we should do is make sure it's not accessible in arith.cpp/.h. I think what I recommended to you Ryan was an address type, but it makes sense that it wouldn't be helpful without a matching unary op that can take in a tensor index and return an address (can just be an int64_t). As long as it's properly protected in arith.cpp to prevent users from ever using this unary op enum, then I think we can go ahead with it.

naoyam replied:
@rdspring1 Thanks for adding the unary op. Are you also adding an address type?

rdspring1 replied:
Not currently. You mean like `ptrdiff_t` => `class Pointer` in `kir` namespace.

`int64_t` is the data type for size, stride, and indices, so I used that for the pointer address.

naoyam replied:
Yes, that's what I was thinking of.

At this point, we don't directly dereference those taken addresses but just calculate offsets, so just `int64_t` would be probably just fine. We may want to reconsider in the future if more complex pointer arithmetic is involved.

naoyam committed:
Do not add kir::TensorIndex::root_domain_index_map_ (#777)

* Do not add kir::TensorIndex::root_domain_index_map_

rdspring1 committed:
Create Address Unary Op

naoyam commented on torch/csrc/jit/codegen/cuda/executor_utils.cpp:
```diff
@@ -163,6 +163,70 @@ bool validateKernelArg(
 
+// Return true if all the tensors have the same stride
+bool checkSameStride(const std::vector<c10::IValue>& tensors) {
+  for (size_t idx = 0; idx < tensors.size() - 1; ++idx) {
```
This seems to be potentially erroneous. Please make sure `tensors.size()` is greater than zero.

naoyam replied:
Actually, C++ tests seem to be failing due to this. I'll fix this.

naoyam replied:
Fixed by 9630219aed14c23ac80a02097d0dca65ac6bb50c

Commit:
Fix unsigned arithmetic

naoyam commented:
I'm seeing 17 C++ tests failing:

```
[  PASSED  ] 252 tests.
[  FAILED  ] 17 tests, listed below:
[  FAILED  ] NVFuserTest.FusionUnaryOps_CUDA
[  FAILED  ] NVFuserTest.FusionSoftmax1D_CUDA
[  FAILED  ] NVFuserTest.FusionSoftmax1DNormalized_CUDA
[  FAILED  ] NVFuserTest.FusionSoftmax3D_CUDA
[  FAILED  ] NVFuserTest.FusionSoftmax3DNormalized_CUDA
[  FAILED  ] NVFuserTest.FusionSmemDynamicPersistentSoftmax2D_CUDA
[  FAILED  ] NVFuserTest.FusionMagicSchedulerSoftmax_CUDA
[  FAILED  ] NVFuserTest.FusionMagicSchedulerLayerNormBackward_CUDA
[  FAILED  ] NVFuserTest.FusionMagicSchedulerLayerNormalization_CUDA
[  FAILED  ] NVFuserTest.FusionMagicSchedulerBatchNormalization_CUDA
[  FAILED  ] NVFuserTest.FusionSmemDynamicPersistentNorm_CUDA
[  FAILED  ] NVFuserTest.FusionThreadPredicate_CUDA
[  FAILED  ] NVFuserTest.FusionLSTMCell_CUDA
[  FAILED  ] NVFuserTest.FusionBiasGeluFwd_CUDA
[  FAILED  ] NVFuserTest.FusionBiasGeluBwd_CUDA
[  FAILED  ] NVFuserTest.FusionSoftmax3DTransposed_CUDA
[  FAILED  ] NVFuserTest.FusionSegmentReduceSoftmax_CUDA

17 FAILED TESTS
```

@rdspring1 , Are we on the same page?

rdspring1 commented:
All the tests pass for me.

```
[----------] Global test environment tear-down
[==========] 259 tests from 1 test case ran. (153661 ms total)
[  PASSED  ] 259 tests.
```

naoyam commented:
That's strange. With my debug build, everything seems fine at de1c1d3, except for the validation bug fixed with the last commit.

naoyam commented:
Sorry, I must have done something wrong. Tried again, seems working now.

naoyam committed:
Clarifies the extent representation of kir::ForLoop (#784)

naoyam commented:
I created another PR to this PR (#788). I'll approve this PR once it's done.

There are still several things that I think should be revisited. For example, adding the `shift` field to `kir::ForLoop` seems too specific to the misaligned vectorization logic. `kir::ForLoop` now has `extent` as well, so as a representation of for-loop, it seems more natural to have `start` and `extent` rather than `extent` and `shift`. 

Another thing I'm not very comfortable with is updating parallel types during indexing. It took me a while why that's needed. I'll revisit there as well in a later PR.

rdspring1 commented:
@csarofeen 

> We want a fixed extent for the for-loop to infer the thread bindings. Instead, we have the `shift` field in the `kir::ForLoop`

naoyam commented:
> @csarofeen
> 
> > We want a fixed extent for the for-loop to infer the thread bindings. Instead, we have the `shift` field in the `kir::ForLoop`

Not exactly sure what you mean, but `shift` and `extent` can be translated to `start = shift`. As an IR, it would be more natural to have `start` and `extent`.

naoyam committed:
Misc cleanups (#788)

* Misc cleanups

csarofeen commented:
Agreed with `start` and `extent`

rdspring1 committed:
Guard against nullptr predicates

naoyam commented:
@rdspring1 Are you merging this PR soon or still adding something? I have some upcoming changes that also modifies looping and indexing logic, so I wonder if you could merge this soon.

rdspring1 commented:
I'm fixing a corner case related to tensor contiguity. It shouldn't take long, but I can merge the commit and create the fix separately.

naoyam commented:
That sounds good to me. I'm not in hurry.

rdspring1 committed:
Restrict merging non-contiguous root domains for misaligned vectorization

rdspring1 committed:
Add check if TensorDomain contains vectorized domain

Allow non-vectorized set ops in vectorized misaligned section

rdspring1 committed:
Add Symbolic Indexing tests

rdspring1 committed:
Update contiguity check

rdspring1 committed:
Use where to set shift to zero correctly

Avoid incorrect result with broadcast tensors in contiguous case

rdspring1 committed:
Update Test Names

csarofeen commented:
@rdspring1 @naoyam looks like there were quite a few commits since the last review, have these changes been reviewed?

rdspring1 commented:
I just added these minor fixes without review.
1. Fix when predicate is a nullptr
2. Use where clause - `(shift == vec_size) ? 0 : shift`
3. Add check to prevent merging non-contiguous root domains

I rebased #819, since I added the same comparison operators.

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+
+  auto pred = PredicateCompute::getInlinePredicate(
+      vec_expr, loop_structure, nullptr, false, true);
+  if (pred == nullptr) {
```
Why is getInlinePredicate returning nullptr?

rdspring1 replied:
I'm passing `nullptr` for the `thread_pred` argument. getInlinePredicate can return 'thread_pred' by itself.
If `thread_pred` isn't `nullptr`, then it will be added to the predicate.

csarofeen replied:
Provide `thread_pred` as `true`. `getInlinePredicate` should not return a `nullptr`

rdspring1 replied:
Then, why do we have a check like this?
https://github.com/csarofeen/pytorch/blob/20_12_3_devel/torch/csrc/jit/codegen/cuda/predicate_compute.cpp#L251-L254

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_validation.cpp:
```diff
@@ -88,8 +88,78 @@ void validateIr(Fusion* fusion) {
 namespace {
 
+// Check contiguity for all root domains associated with Misaligned Vectorize ParallelType
+void checkContiguity(
```
This should be `assertFullContiguity` not `checkContiguity`

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_validation.cpp:
```diff
@@ -88,8 +88,78 @@ void validateIr(Fusion* fusion) {
+
+  std::unordered_map<IterDomain*, bool> producer_domain_contiguity;
+  for (size_t idx = 0; idx < producer->getRootDomain().size(); ++idx) {
+    auto root = producer->getRootDomain()[idx];
```
`root_id` not `root`

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_validation.cpp:
```diff
@@ -88,8 +88,78 @@ void validateIr(Fusion* fusion) {
+    producer_domain_contiguity.insert({root, contiguity});
+  }
+
+  for (auto consumer_root : consumer->getRootDomain()) {
```
`consumer_root_id` not `consumer_root`

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_validation.cpp:
```diff
@@ -88,8 +88,78 @@ void validateIr(Fusion* fusion) {
+}
+
+// Check contiguity for all root domains associated with Misaligned Vectorize ParallelType
+void checkContiguity(
```
`assertMappedConsumerContiguity`

csarofeen replied:
I don't know what property you're checking for here, so I really don't know if this is doing what you expect.

rdspring1 replied:
We cannot merge any non-contiguous root domains for misaligned vectorization (MV).
This function checks if all the root domains associated with MV iter_domain are contiguous.

We apply the MV parallel type to the local TensorView, but the contiguity information is stored with the global TensorView. This function uses the PairwiseRootDomainMap to map the root domains between the TensorViews.

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_validation.cpp:
```diff
@@ -109,16 +181,20 @@ class VectorizeValidator : public OptInDispatch {
   }
 
  private:
+  std::unordered_set<IterDomain*> domains_;
```
What is this member? Domains of what?

rdspring1 replied:
This is all the iter_domains associated with the vectorized iter_domain. I then filter all the iter_domains to get only the root domains.

naoyam replied:
I think you can do `IterVsitor::getInputsTo(vectorized_id)`, which should return only root IDs that are an input to `vectorized_id`.

naoyam replied:
It's traversing the expression list anyway, so it may actually make sense not to do additional traversal with `getInputsTo`. In any case, adding `s->outer()` and `s->inner()` in the split case should not be necessary, right?

rdspring1 replied:
Why isn't adding either necessary?

naoyam replied:
If I understand it correctly, it is intended to find root axes by traversing backwards, right? The outer and inner of a `split` are output, so why are they needed to be added? It doesn't make the end result wrong as only root axes are used from the final resulting set, but I don't understand why they need to be added.

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+// for (index = start; index < extent; index += offset)
+// vectorize flag - Do not generate for-loop
+// shift value - Add shift to global indices generated within For-Loop
+void cloneVectorizeLoopNests(
```
How does a clone function not return anything?

csarofeen replied:
It's not clear from reading this code what this function does.

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+             ->domain()
+             ->hasVectorize());
+
+    TORCH_INTERNAL_ASSERT(!has_vectorize_op || fl->body().exprs().size() == 1);
```
Commenting why this is the case would be helpful.

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+
+// Find any child For-Loops
+// Add remaining expressions to new parent For-Loop
+std::vector<kir::ForLoop*> parseVectorizedForLoop(
```
What is this doing? Need more descriptive comments and some context.

rdspring1 replied:
This function separates the allocations from the child for-loops. We need to duplicate the child-for loops. I see no reason to duplicate the allocations, so I move them to the top of the for-loop.

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+// Enable vectorize flag in child For-Loop
+kir::Expr* findVectorizedSet(
+    std::vector<kir::ForLoop*>& loop_structure,
+    const std::vector<kir::ForLoop*>& for_loops) {
```
What's the difference between `loop_structure` and `for_loops` this is vague.

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+
+  TORCH_INTERNAL_ASSERT(domain.size() == indices.size());
+
+  bool is_contiguous = true;
```
fully contiguous

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+    is_contiguous &= status;
+  }
+
+  // If the tensorview is not contiguous, return inner-most root domain extent
```
Clearly state assumptions on other parts of the code base. This isn't a generic thing, this is enforced in other parts of the PR.

rdspring1 replied:
I'm not sure what you mean here.

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+      // vectorization partition.
+      extent = ir_builder.mulExpr(extent, root_id->extent());
+    } else {
+      break;
```
Unclear to me why we'd hit break from reading this code. Add comments.

rdspring1 replied:
@naoyam why is there a break here?

naoyam replied:
The loop here finds the maximum consecutive root axes that are contiguous. For global indexing, `indices.at(i)` being zero means it's contiguous, so that's what I used here. However, I don't understand why I did that rather than just using the contiguity vector. I'll clean up this part.

naoyam replied:
See #826

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+  auto alloc = ir_builder.create<kir::Allocate>(
+      namedScalar, MemoryType::Local, ir_builder.one());
+  body.push_back(alloc);
+  body.push_back(namedScalar->definition());
```
Is this guaranteed to not be nullptr?

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+  return namedScalar;
+}
+
+kir::ForLoop* handleMisalignedVectorization(
```
What does handle mean?

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+
+kir::ForLoop* handleMisalignedVectorization(
+    std::vector<kir::ForLoop*> loop_structure,
+    const kir::ForLoop* for_loop) {
```
what's the difference between `loop_structure` and `for_loop`?

rdspring1 replied:
`loop_structure` corresponds with https://github.com/csarofeen/pytorch/blob/20_12_3_devel/torch/csrc/jit/codegen/cuda/lower_unroll.h#L86-L87

The `for_loop` is the loop that contains the misaligned vectorization ops.

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+kir::ForLoop* handleMisalignedVectorization(
+    std::vector<kir::ForLoop*> loop_structure,
+    const kir::ForLoop* for_loop) {
+  // for_loop body contains allocate, read, compute, write operations
```
Why? What is it?

rdspring1 replied:
The base for-loop contains the for-loops with the vectorized sets.
```
// base for-loop 
for (....) {
  // Allocations
  for (...) {
    // Read - UnaryOp::Set - Vectorize
  }
 // Allocations
  for (...) {
     // Computation
  }
 // Allocations
  for (...) {
    // Write - UnaryOp::Set - Vectorize
  }
}
```

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+  // for_loop body contains allocate, read, compute, write operations
+  kir::IrBuilder ir_builder(GpuLower::current()->kernel());
+
+  // create new base For-Loop
```
What's a base for loop?

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+  auto child_loops = parseVectorizedForLoop(for_loop, new_loop);
+
+  // Find the first vectorize set - either read or write
+  auto vec_expr = findVectorizedSet(loop_structure, child_loops);
```
I assume this is a `set` like "UnaryOp::Set` not `unordered_set`?

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+  auto in_tv = vec_expr->inputs().front()->as<kir::TensorView>();
+
+  // It is assumed that either of input and output is on global memory
+  TORCH_INTERNAL_ASSERT(
```
This seems overly redundant, one has to be global and the other has to be local, can't we check for that in one assert?

csarofeen replied:
This should be an interface check, though could also be asserted here.

rdspring1 replied:
@naoyam Couldn't we merge these assertions into a single check?

naoyam replied:
Please feel free to make it simpler. Agree I was overly redundant. I wanted document the assumptions explicitly.

However, please do not remove the assertions from this function entirely. I know they are already validated, but I believe it's safer to have these assertions here in case when we attempt to remove those assumptions.

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+      vec_tv->domain()->domain().back()->extent()->as<kir::Int>();
+
+  auto index =
+      ir_builder.create<kir::TensorIndex>(global_tv->fuserTv(), indices);
```
We should not be doing the direct indexing in an unrolling pass. This is taking a step in the wrong direction. Quite honestly we should be getting predicate computation out of unrolling.

rdspring1 replied:
This is the approach you proposed in February instead of using only for-loops.

From 2/8/2021:
> RS: why do you want to go with this approach?
CS: it has a fixed extent on the for loop
CS: otherwise good luck inferring your thread bindings

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+
+  auto shift_pred = ir_builder.eqExpr(shift_init, vector_size);
+  auto shift_val =
+      ir_builder.whereExpr(shift_pred, ir_builder.zero(), shift_init);
```
I'd really like to understand why we're using a `whereExpr` here, this seems odd to me.

rdspring1 replied:
`shift = vec_size - (address % vec_size)`
The range of shift is [1, vec_size]. However, this can cause an issue with broadcast tensors.
The where clause maps the range to [0, vec_size-1].
`shift = (shift == vec_size) ? 0 : shift`

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -32,6 +32,281 @@ kir::ForLoop* cloneLoopNest(const kir::ForLoop* for_loop, bool unroll = false) {
+  //  Get full extent for merged root domains
+  auto extent = getVectorizeExtent(vec_tv, indices);
+
+  auto vector_size =
```
Need lots of comments in the following code.

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -54,6 +329,19 @@ bool isReductionInitExpr(const kir::Expr* expr) {
   return true;
 }
 
+bool containsMisalignedVectorization(const kir::ForLoop* fl) {
```
From the name I would expect this to be recursive, what you're really doing is seeing if a direct child of `fl` is set as misaligned vectorize.

rdspring1 replied:
I want the for-loop that contains vectorized UnaryOpType::Set operations.
Reads and writes use the same shift value.

csarofeen reviewed (commented):
CC @naoyam

naoyam commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -110,13 +110,25 @@ kir::Expr* findVectorizedSet(
+
+  // If the tensorview is not contiguous, return inner-most root domain extent
+  if (!is_contiguous) {
+    return domain.back()->extent();
```
Why is this necessary? Do you have a test case that fails without this?

rdspring1 replied:
We cannot merge non-contiguous root domains. Otherwise, we can get incorrect results.

2D Non-Contiguous Tensor Example:
```
[     row 0 - shift 0  ]       [     row 1 - shift 1  ]
[pre | vectorize | post] [***] [pre | vectorize | post]
                     ^-----------^ overlapping rows
```

If we merge non-contiguous root domains, we can index into the post and pre section of two different rows.
This would require two shift calculations, which isn't currently supported.

It is related to `FusionVectorizeMisalignedPointwiseMergeSymbolicPass_CUDA` and `FusionVectorizeMisalignedPointwiseMergeSymbolicFail_CUDA`.

naoyam replied:
It only extends the vectorized domain when the index into a root axis is zero. IIUC, that happens iff the root axis is contiguous. I may be missing something, so I'm curious when it could fail without the added check. I commented out the `is_contiguous` part and didn't see any failure.

rdspring1 replied:
Why is it necessary for there to be a failure to add this to the function?
I didn't see a reason to finish the function in this scenario.

naoyam replied:
Sorry, I don't understand what you mean. Can you please rephrase your comment?

naoyam replied:
See #826

naoyam commented on torch/csrc/jit/codegen/cuda/lower_validation.cpp:
```diff
@@ -88,8 +88,78 @@ void validateIr(Fusion* fusion) {
+          consumer);
+
+      TORCH_INTERNAL_ASSERT(
+          root_c2p.find(consumer_root) != root_c2p.end());
```
This should be moved up before `auto producer_root = root_c2p[consumer_root];`.

rdspring1 replied:
👍🏼

naoyam commented on torch/csrc/jit/codegen/cuda/lower_validation.cpp:
```diff
@@ -88,8 +88,78 @@ void validateIr(Fusion* fusion) {
+          producer_domain_contiguity.find(producer_root) !=
+          producer_domain_contiguity.end());
+
+      TORCH_INTERNAL_ASSERT(
+          !consumer_root->isBroadcast() || !producer_root->isBroadcast(),
+          "Misaligned vectorization prohibits merging broadcast domains.",
+          "Issue found in, ",
+          consumer);
```
This isn't really checking contiguity. It should be done in a separate function, or the name of this function should be renamed (just `validateMisalignedVectoizationRootDomain` or something would be less confusing).

rdspring1 replied:
I'll rename the function.

naoyam commented on torch/csrc/jit/codegen/cuda/lower_validation.cpp:
```diff
@@ -181,6 +258,20 @@ class VectorizeValidator : public OptInDispatch {
+      if (tv->getMemoryType() == MemoryType::Global) {
+        checkContiguity(validator.domains_, tv);
+      } else if (
+          tv->definition()->getExprType() == ExprType::UnaryOp &&
+          tv->definition()->as<UnaryOp>()->getUnaryOpType() ==
+              UnaryOpType::Set) {
```
Shouldn't it just fail if these condition were not met?

rdspring1 replied:
👍🏼

naoyam commented on torch/csrc/jit/codegen/cuda/kernel_ir.cpp:
```diff
@@ -167,6 +167,13 @@ bool TensorDomain::hasRFactor() const {
 
+bool TensorDomain::hasVectorize() const {
+  return std::any_of(domain_.begin(), domain_.end(), [](IterDomain* id) {
+    return id->parallelType() == ParallelType::Vectorize ||
+        id->parallelType() == ParallelType::MisalignedVectorize;
```
Please use `isParallelTypeVectorize` in type.h.

naoyam commented on torch/csrc/jit/codegen/cuda/ir_nodes.cpp:
```diff
@@ -896,6 +896,13 @@ bool TensorDomain::hasRFactor() const {
 
+bool TensorDomain::hasVectorize() const {
+  return std::any_of(domain_.begin(), domain_.end(), [](IterDomain* id) {
+    return id->getParallelType() == ParallelType::Vectorize ||
+        id->getParallelType() == ParallelType::MisalignedVectorize;
```
Please use `isParallelTypeVectorize` in type.h.

naoyam commented on torch/csrc/jit/codegen/cuda/lower_unroll.cpp:
```diff
@@ -49,7 +49,9 @@ void cloneVectorizeLoopNests(
-         first_expr->as<kir::UnaryOp>()->operation() == UnaryOpType::Set);
+         first_expr->as<kir::UnaryOp>()->operation() == UnaryOpType::Set &&
+         first_expr->as<kir::UnaryOp>()->out()->isA<kir::TensorView>() &&
+         first_expr->as<kir::UnaryOp>()->out()->as<kir::TensorView>()->domain()->hasVectorize());
```
Having a utility function like `isVectorizedSet(Expr*)` would be helpful.

rdspring1 replied:
👍🏼

naoyam commented on torch/csrc/jit/codegen/cuda/runtime/helpers.cu:
```diff
@@ -85,6 +85,10 @@ __device__ float where(bool c, float a, float b) {
 }
 
+__device__ float where(bool c, int64_t a, int64_t b) {
+  return c ? a : b;
```
Shouldn't the return type be `int64_t`?

naoyam replied:
Maybe, making it a template?

rdspring1 replied:
Using a template function make sense.

naoyam commented on torch/csrc/jit/codegen/cuda/executor_utils.cpp:
```diff
@@ -193,23 +193,25 @@ bool checkSameStride(const std::vector<c10::IValue>& tensors) {
 }
 
 // Return true if all the tensors have the same stride
```
-> `Return true if all the tensors have the same stride and contiguous.`

naoyam commented:
I looked at the commits made after my last review and added some comments.

naoyam commented:
@rdspring1 Just a reminder there're several questions and suggestions that I'm waiting for your response.

rdspring1 commented:
I responded to your comments.
